pkg/kubelet/nodeshutdown/nodeshutdown_manager_linux.go

```diff
@@ -0,0 +1,232 @@
+		if err != nil {
+			return err
+		}
+
+		err = m.dbusCon.ReloadLogindConf()
+		if err != nil {
+			return err
+		}
+
+		// Read the current inhibitDelay again, if the override was successful, currentInhibitDelay will be equal to shutdownGracePeriodRequested.
+		updatedInhibitDelay, err := m.dbusCon.CurrentInhibitDelay()
+		if err != nil {
+			return err
+		}
+
+		if updatedInhibitDelay != m.shutdownGracePeriodRequested {
+			return fmt.Errorf("node shutdown manager was unable to update logind InhibitDelayMaxSec to %v (ShutdownGracePeriod), current value of InhibitDelayMaxSec (%v) is less than requested ShutdownGracePeriod", m.shutdownGracePeriodRequested, updatedInhibitDelay)
+		}
+	}
+
+	err = m.aquireInhibitLock()
+	if err != nil {
+		return err
+	}
+
+	events, err := m.dbusCon.MonitorShutdown()
+	if err != nil {
+		return err
+	}
+
+	go func() {
+		for {
+			select {
+			case isShuttingDown := <-events:
+				klog.V(1).Infof("Shutdown manager detected new shutdown event, isNodeShuttingDownNow: %t", isShuttingDown)
+				m.nodeShuttingDownMutex.Lock()
+				m.nodeShuttingDownNow = isShuttingDown
+				m.nodeShuttingDownMutex.Unlock()
+				if isShuttingDown {
+					m.processShutdownEvent()
+				} else {
+					m.aquireInhibitLock()
+				}
+			}
+		}
+	}()
+	return nil
+}
+
+func (m *Manager) aquireInhibitLock() error {
+	lock, err := m.dbusCon.InhibitShutdown()
+	if err != nil {
+		return err
+	}
+	m.inhibitLock = lock
+	return nil
+}
+
+// ShutdownStatus will return an error if the node is currently shutting down.
+func (m *Manager) ShutdownStatus() error {
+	if !utilfeature.DefaultFeatureGate.Enabled(features.GracefulNodeShutdown) {
+		return nil
+	}
+
+	m.nodeShuttingDownMutex.Lock()
+	defer m.nodeShuttingDownMutex.Unlock()
+
+	if m.nodeShuttingDownNow {
+		return fmt.Errorf("node is shutting down")
+	}
+	return nil
+}
+
+func (m *Manager) processShutdownEvent() error {
+	klog.V(1).Infof("Shutdown manager processing shutdown event")
+	activePods := m.getPods()
+
+	nonCriticalPodGracePeriod := m.shutdownGracePeriodRequested - m.shutdownGracePeriodCriticalPods
+
+	var wg sync.WaitGroup
+	wg.Add(len(activePods))
+	for _, pod := range activePods {
+		go func(pod *v1.Pod) {
+			defer wg.Done()
+
+			var gracePeriodOverride int64
+			if kubelettypes.IsCriticalPod(pod) {
+				gracePeriodOverride = int64(m.shutdownGracePeriodCriticalPods.Seconds())
+				m.clock.Sleep(nonCriticalPodGracePeriod)
+			} else {
+				gracePeriodOverride = int64(nonCriticalPodGracePeriod.Seconds())
+			}
+
+			// If the pod's spec specifies a termination gracePeriod which is less than the gracePeriodOverride calculated, use the pod spec termination gracePeriod.
+			if pod.Spec.TerminationGracePeriodSeconds != nil && *pod.Spec.TerminationGracePeriodSeconds <= gracePeriodOverride {
+				gracePeriodOverride = *pod.Spec.TerminationGracePeriodSeconds
+			}
+
+			klog.V(1).Infof("Shutdown manager killing pod %q with gracePeriod: %v", pod.Name, gracePeriodOverride)
```

Comment:
specifying units might be helpful:

```suggestion
			klog.V(1).Infof("Shutdown manager killing pod %q with gracePeriod %v seconds", pod.Name, gracePeriodOverride)
```

Reply:
good idea, done.